dali/operators/math/expressions/expression_factory_instances/expression_impl_factory.h

```diff
@@ -105,6 +105,60 @@ std::unique_ptr<ExprImplBase> ExprImplFactoryBinOp(const ExprFunc &expr) {
   return result;
 }
 
+
+/**
+ * @brief Inspect `expr` to transform runtime information to static information, obtain the
+ *        implementation for ternary op (executor for given expression) and return it.
+ *
+ * The static type switch goes over input types and input kinds.
+ * This is ternary case and seven possible input kind combinations are supported,
+ * that is all combinations of Tensors and Constants, except (Constant, Constant, Constant)
+ *
+ * @tparam Impl template that maps binary Arithmetic Op and input/output types
+ *                          to a functor that can execute it over a tile of three tensors.
+ *                          The implementation need to loop over inputs and output.
+ *                          @see ExprImplCpuTernary or ExprImplGpuTernary
+ * @param expr The expression node for which we return the executor
+ */
+template <ArithmeticOp op, typename Backend,
+          template <ArithmeticOp, typename, typename, typename, typename, bool, bool, bool>
+          class ImplsAll>
+std::unique_ptr<ExprImplBase> ExprImplFactoryTernaryOp(const ExprFunc &expr) {
+  std::unique_ptr<ExprImplBase> result;
+  DALIDataType types[3];
+  bool is_tensor[3];
+  for (int i = 0; i < 3; i++) {
+    types[i] = expr[i].GetTypeId();
+    is_tensor[i] = !IsScalarLike(expr[i]);
+  }
+
+  TYPE_SWITCH(types[0], type2id, First, ARITHMETIC_ALLOWED_TYPES, (
+    TYPE_SWITCH(types[1], type2id, Second, ARITHMETIC_ALLOWED_TYPES, (
+      TYPE_SWITCH(types[2], type2id, Third, ARITHMETIC_ALLOWED_TYPES, (
+        using Out_t = typename arithm_meta<op, Backend>::template result_t<First, Second, Third>;
+        if (is_tensor[0] && is_tensor[1] && is_tensor[2]) {
```

Comment:
I wonder if nested if wouldn't be easier to read:
```
else if (is_tensor[0]) {
  if (is_tensor[1]) {
     if (!is_tensor[2]) {
  else { // !is_tensor[1
    if (!is_tensor[2]) {
    else { // !is_tensor[2]
```

Reply:
I'm not sure, that would be probably more complicated and more nested.

Reply:
Ok